impl/src/main/java/org/jboss/cdi/tck/tests/beanContainer/assignability/BeanObserverAssignabilityTest.java

```diff
@@ -0,0 +1,187 @@
+
+import static org.jboss.cdi.tck.cdi.Sections.BM_BEAN_EVENT_ASSIGNABILITY;
+import static org.testng.Assert.assertFalse;
+import static org.testng.Assert.assertThrows;
+import static org.testng.Assert.assertTrue;
+
+import java.lang.reflect.Type;
+import java.util.Set;
+
+import org.jboss.arquillian.container.test.api.Deployment;
+import org.jboss.cdi.tck.AbstractTest;
+import org.jboss.cdi.tck.cdi.Sections;
+import org.jboss.cdi.tck.shrinkwrap.WebArchiveBuilder;
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+import org.jboss.test.audit.annotations.SpecAssertion;
+import org.jboss.test.audit.annotations.SpecVersion;
+import org.testng.annotations.Test;
+
+import jakarta.enterprise.inject.Any;
+import jakarta.enterprise.inject.Default;
+import jakarta.enterprise.inject.literal.NamedLiteral;
+
+/**
+ * @author Andrew Rouse
+ */
+@SpecVersion(spec = "cdi", version = "4.1")
+public class BeanObserverAssignabilityTest extends AbstractTest {
+
+    @Deployment
+    public static WebArchive createTestArchive() {
+        return new WebArchiveBuilder().withTestClassPackage(BeanObserverAssignabilityTest.class).build();
+    }
+
+    @Test
+    @SpecAssertion(section = BM_BEAN_EVENT_ASSIGNABILITY, id = "aa")
+    public void testBeanMatching() {
+        Set<Type> beanTypes = Set.of(MyBean.class, MyBeanInterface.class, Object.class);
+
+        // Basic checks with qualifiers
+        assertTrue(getCurrentBeanContainer().isMatchingBean(beanTypes, Set.of(), MyBean.class, Set.of()),
+                "Bean did not match its own type");
+        assertFalse(getCurrentBeanContainer().isMatchingBean(beanTypes, Set.of(), MyBean.class, Set.of(Qualifier1.Literal.INSTANCE)),
+                "Bean matched despite not having required qualifier");
+        assertTrue(getCurrentBeanContainer().isMatchingBean(beanTypes, Set.of(Qualifier1.Literal.INSTANCE), MyBean.class, Set.of(Qualifier1.Literal.INSTANCE)),
+                "Bean did not match despite having required qualifier");
+        assertTrue(getCurrentBeanContainer().isMatchingBean(beanTypes, Set.of(Qualifier1.Literal.INSTANCE, Qualifier2.Literal.INSTANCE), MyBean.class, Set.of(Qualifier1.Literal.INSTANCE)),
+                "Bean did not match despite having a superset of the required qualifiers");
+        
+        // Only bean types passed in should be considered
+        Set<Type> reducedBeanTypes = Set.of(MyBean.class);
+        assertTrue(getCurrentBeanContainer().isMatchingBean(reducedBeanTypes, Set.of(), MyBean.class, Set.of()),
+                "Bean did not match its own type");
+        assertFalse(getCurrentBeanContainer().isMatchingBean(reducedBeanTypes, Set.of(), MyBeanInterface.class, Set.of()),
+                "Bean matched MyBeanInterface despite it not being in bean types");
+        assertFalse(getCurrentBeanContainer().isMatchingBean(reducedBeanTypes, Set.of(), Object.class, Set.of()),
+                "Bean matched Object despite it not being in bean types");
```

Comment:
Wasn't sure about this test.

Every bean has `Object` as a bean type. Should `Object` be implied if not passed in the `beanTypes` set?

Reply:
> Every bean has `Object` as a bean type. Should `Object` be implied if not passed in the `beanTypes` set?

Yes, that's always added.
I think there was even a TCK for `@Typed` that asserted that `Object` is added automatically.

Reply:
(Lurking. Curious if this is generally true in the specification? I see that certainly for `@Typed` `Object.class` is implied, but if I make a synthetic bean and explicitly do not include `Object.class` in its `Set<Type>`...?)

Reply:
> For example, https://jakarta.ee/specifications/cdi/4.0/apidocs/jakarta.cdi/jakarta/enterprise/inject/spi/configurator/beanattributesconfigurator#types(java.util.Set) says nothing about Object.class being intrinsically added.)

Yes, I am aware of that. However, event a synth bean is a bean and as therefore the first link you posted should still apply. At least in my opinion :)
And I think that's also what Weld does; not sure if there is a TCK for it though 🤷

Reply:
It might be helpful to clarify whether `Object` is automatically added with a note on the relevant methods.

For this method, it sounds like we think `Object` _should_ be automatically added to the set of bean types so this test should change to `assertTrue`?

Reply:
If all synthetic beans are beans, and if all beans have the bean type `java.lang.Object`, then all synthetic beans have the bean type `java.lang.Object`. That is a straightforward implication in my eyes.

Now, whether `isMatchingBean` should automatically add `Object` to `beanTypes`, that's another question. It could, I don't think that would be harmful in any way, but the method specification should be explicit about it. Anyone wants to submit a PR?

The only possible downside is that someone might start asking "why don't we automatically scan for all superclasses and implemented interfaces?", and I wouldn't have a good answer :-)

Reply:
Here's a proposal: https://github.com/jakartaee/cdi/pull/744 I went full on the precise mode, as that's just easiest frankly :-)

Reply:
> their own implementation of `Bean` which may not return `Object` from `getTypes`.

That would be an invalid implementation of `Bean` in my eyes, though the container can compensate for it. All the more reason for using the `BeanConfigurator` API.

Reply:
> Here's a proposal

That looks good, thank you :-)

I've updated the test to match.

> That would be an invalid implementation of `Bean` in my eyes

Yeah, I agree and it's not the only way you could have an invalid `Bean` implementation either (e.g. returning a non-scope annotation from `getScope()`). I guess we could be more specific about how and when these sorts of errors are handled.

I note that `BeanConfigurator` is still vulnerable to some of the same issues. You could fail to pass `Object` to [`types`][types] as @ljnelson suggested, or pass a non-scope annotation to [`scope`][scope] and we don't currently document what should happen.

[types]: https://jakarta.ee/specifications/cdi/4.0/apidocs/jakarta.cdi/jakarta/enterprise/inject/spi/configurator/beanattributesconfigurator#types(java.lang.reflect.Type...)
[scope]: https://jakarta.ee/specifications/cdi/4.0/apidocs/jakarta.cdi/jakarta/enterprise/inject/spi/configurator/beanattributesconfigurator#scope(java.lang.Class)

Reply:
Good points. Out of scope of this PR for sure, though :-)

Reply:
> If all synthetic beans are beans, and if all beans have the bean type java.lang.Object, then all synthetic beans have the bean type java.lang.Object. That is a straightforward implication in my eyes.

I am of the same opinion.

> The only possible downside is that someone might start asking "why don't we automatically scan for all superclasses and implemented interfaces?", and I wouldn't have a good answer :-)

Because you might want to simulate a scenario where `@Typed` was used and some types might be omitted.

> Yeah, I agree and it's not the only way you could have an invalid Bean implementation either (e.g. returning a non-scope annotation from getScope()). I guess we could be more specific about how and when these sorts of errors are handled.

Well, since you can define arbitrary things if you implement `Bean` yourself, it would be rather extensive to nail it all down...